### ISSUE TITLE
%matplotlib notebook issue

### ISSUE THREAD

**Issue by Zhilei** [hpc4cmb/toast-workshop-ucsd-2019#21]
I was trying to use
`%matplotlib notebook` instead of `%matplotlib inline` to have interactive plotting tools. But I got errors with the last line as 

> ModuleNotFoundError: No module named 'ipympl'

I was wondering whether this can be fixed by modifying the loaded modules.

Thank you so much!

**Comment by ickc** [comment]
Can you install that dep. by yourself? I think the way Ted set it up point the user packages location somewhere in your home. So in the terminal if you load the cmbenv and conda/pip install then probably it will work.

**Comment by tskisner** [comment]
I just installed this in the live version (20191012) and added it to the default list of packages:

https://github.com/hpc4cmb/cmbenv/commit/67505cf87b872d86bc80cf3e7b969f5f0079eb3d

As @ickc mentioned, one could also probably pip install it to your user directory.  Just be careful if you are swapping between incompatible python stacks and trying to use a package installed in your user space.  One could also create a new conda env, but then you would have to replicate all the packages in the cmbenv stack.

**Comment by zonca** [comment]
@tskisner is this fixed in the cmbenv we are going to use for Trieste?

**Comment by tskisner** [comment]
I did install the package requested, however, I think using this feature requires support from the NERSC jupyter installation, not our software stack.

**Comment by zonca** [comment]
I looked at the docs at https://github.com/matplotlib/jupyter-matplotlib,
it looks like to have JupyterLab support we need:

```
# If using JupyterLab
conda install nodejs
jupyter labextension install @jupyter-widgets/jupyterlab-manager
jupyter labextension install jupyter-matplotlib
```

However it is quite annoying to have to install `nodejs`...
@tskisner If you think this is easy enough it would be good to have, otherwise let's close this issue.

**Comment by ickc** [comment]
Since the Jupyterlab instance is not started by us, but provided at NERSC, I don't think we can install our own jupyter lab extension? I think this is what @tskisner is referring to.

I wonder if [spin](https://www.nersc.gov/users/data-analytics/spin/) can be used to deploy our own stack including the Jupyterlab itself, and I wonder if MPI can even be used with spin. But Spin has been in the "pilot phase" for a long time meaning that only people registered for training can deploy spin applications. (Spin relies on container.)

**Comment by zonca** [comment]
ah yes, sorry,
then I'll open a ticket and ask if they can install this

**Comment by tskisner** [comment]
Yes, I was saying exactly what @ickc mentioned- this is something that would have to be changed by NERSC.  The cmbenv stack already contains jupyter-notebook, so I also tried this:
```
$>  module load cmbenv
$>  source cmbenv
$>  salloc -N 1 -C haswell -A mp107 --qos=interactive -t 04:00:00
%> hostname
     nid00077
%> ifconfig -a
note IP address = 10.128.0.78
%> jupyter-notebook --no-browser --port 8889
```
And then I installed firefox on cori and tried to connect to my running notebook on the compute node using the IP address.  However, it looks like they are blocking direct connections.

**Comment by zonca** [comment]
Rollin replied that the extension is already installed, in fact it works with the default kernel (but using `%maptlotlib widget` instead of `%matplotlib notebook`

![image](https://user-images.githubusercontent.com/383090/68168314-3d5ab500-ff1d-11e9-816a-71d771e74b8e.png)

Now if I test with CMB 20191012, also using `%matplotlib widget`, it hangs at "Loading widget" my guess is that there are different versions of some of the libraries between our kernel and the NERSC Kernel.

![image](https://user-images.githubusercontent.com/383090/68168559-08029700-ff1e-11e9-8d60-a6ba615d10b4.png)

**Comment by zonca** [comment]
I see we have 0.2.2 and they have 0.3.3, @tskisner can you try to update `ipympl` to `0.3.3`.
If this doesn't work let's close this issue anyway, don't want to spend too much time.

**Comment by ickc** [comment]
> And then I installed firefox on cori and tried to connect to my running notebook on the compute node using the IP address. However, it looks like they are blocking direct connections.

In the past when I tried, it only works when it is reverse tunneling.

**Comment by rcthomas** [comment]
Indeed this is probably a compatibility issue and may be resolved by upgrading ipympl.

**Comment by rcthomas** [comment]
BTW let me know if you guys want to meet to discuss your needs, happy to help.
